WordPress/gutenberg
# Block Editor: LinkControl: Prevent focus loss in edit mode toggle

Previously: https://github.com/WordPress/gutenberg/pull/19885#issuecomment-578967785

This pull request seeks to handle a focus loss which can occur in the `LinkControl` component after choosing a new link value. This occurs because the DOM changes such that the current active element (the link input) is suddenly removed. This is now anticipated by the component, making sure that if it had focused _before_ the state is toggled, focus is restored _after_ the state is toggled if it resulted in a focus loss.

Note: There is some overlap between this and #19856, where as of #19856, it may not be necessarily guaranteed that submitting a link would cause the input to be removed, if the next render does not pass a usable `value`.

**Testing Instructions:**

New test assertions have been added for the component's unit tests, and the end-to-end tests for both the Buttons and Navigation blocks.

```
npm run test-unit packages/block-editor/src/components/link-control/test/index.js
npm run test-e2e packages/e2e-tests/specs/editor/blocks/navigation.test.js
npm run test-e2e packages/e2e-tests/specs/editor/blocks/buttons.test.js
```

Verify there is no focus loss when submitting a link:

1. Navigate to Posts > Add New
2. Insert a Buttons block
3. Click "Link" in the block toolbar (or press <kbd>Cmd+K</kbd>)
4. Input a URL or search term and either press <kbd>Enter</kbd> or select a suggestion
5. Note that the URL of the URL preview should be focused
6. Press <kbd>Escape</kbd>
7. Note that the popover is closed
